// device-uuid/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// src/constants/index.ts
var BOTS = [
  "\\+https:\\/\\/developers.google.com\\/\\+\\/web\\/snippet\\/",
  "googlebot",
  "baiduspider",
  "gurujibot",
  "yandexbot",
  "slurp",
  "msnbot",
  "bingbot",
  "facebookexternalhit",
  "linkedinbot",
  "twitterbot",
  "slackbot",
  "telegrambot",
  "applebot",
  "pingdom",
  "tumblr ",
  "Embedly",
  "spbot"
];
var IS_BOT_REGEXP = new RegExp(`^.*(${BOTS.join("|")}).*$`);
var VERSION_PATTERNS = {
  Edge: /(?:Edge|Edg)\/([\d\w.-]+)/i,
  Firefox: /firefox\/([\d\w.-]+)/i,
  IE: /msie\s([\d.]+[\d])|trident\/\d+\.\d+;.*[rv:]+(\d+\.\d)/i,
  Chrome: /chrome\/([\d\w.-]+)/i,
  Chromium: /(?:chromium|crios)\/([\d\w.-]+)/i,
  Safari: /version\/([\d\w.-]+)/i,
  Opera: /version\/([\d\w.-]+)|OPR\/([\d\w.-]+)/i,
  Ps3: /([\d\w.-]+)\)\s*$/i,
  Psp: /([\d\w.-]+)\)?\s*$/i,
  Amaya: /amaya\/([\d\w.-]+)/i,
  SeaMonkey: /seamonkey\/([\d\w.-]+)/i,
  OmniWeb: /omniweb\/v([\d\w.-]+)/i,
  Flock: /flock\/([\d\w.-]+)/i,
  Epiphany: /epiphany\/([\d\w.-]+)/i,
  WinJs: /msapphost\/([\d\w.-]+)/i,
  PhantomJS: /phantomjs\/([\d\w.-]+)/i,
  UC: /UCBrowser\/([\d\w.]+)/i
};
var BROWSER_PATTERNS = {
  Edge: /edge|edg\//i,
  Amaya: /amaya/i,
  Konqueror: /konqueror/i,
  Epiphany: /epiphany/i,
  SeaMonkey: /seamonkey/i,
  Flock: /flock/i,
  OmniWeb: /omniweb/i,
  Chromium: /chromium|crios/i,
  Chrome: /chrome/i,
  Safari: /safari/i,
  IE: /msie|trident/i,
  Opera: /opera|OPR/i,
  PS3: /playstation 3/i,
  PSP: /playstation portable/i,
  Firefox: /firefox/i,
  WinJs: /msapphost/i,
  PhantomJS: /phantomjs/i,
  UC: /UCBrowser/i
};
var OS_PATTERNS = {
  Windows11: /windows nt 10\.0.*(?:; win64; x64|; wow64).*(?:rv:|edg\/|chrome\/).*?(\d+)/i,
  Windows10: /windows nt 10\.0/i,
  Windows81: /windows nt 6\.3/i,
  Windows8: /windows nt 6\.2/i,
  Windows7: /windows nt 6\.1/i,
  UnknownWindows: /windows nt 6\.\d+/i,
  WindowsVista: /windows nt 6\.0/i,
  Windows2003: /windows nt 5\.2/i,
  WindowsXP: /windows nt 5\.1/i,
  Windows2000: /windows nt 5\.0/i,
  WindowsPhone8: /windows phone 8\./,
  OSXCheetah: /os x 10[._]0/i,
  OSXPuma: /os x 10[._]1(\D|$)/i,
  OSXJaguar: /os x 10[._]2/i,
  OSXPanther: /os x 10[._]3/i,
  OSXTiger: /os x 10[._]4/i,
  OSXLeopard: /os x 10[._]5/i,
  OSXSnowLeopard: /os x 10[._]6/i,
  OSXLion: /os x 10[._]7/i,
  OSXMountainLion: /os x 10[._]8/i,
  OSXMavericks: /os x 10[._]9/i,
  OSXYosemite: /os x 10[._]10/i,
  OSXElCapitan: /os x 10[._]11/i,
  OSXSierra: /os x 10[._]12/i,
  OSXHighSierra: /os x 10[._]13/i,
  OSXMojave: /os x 10[._]14/i,
  OSXCatalina: /os x 10[._]15/i,
  MacOSBigSur: /mac os x 1[01][._](?:16|[6-9]|\d{2})|os x 10[._]16/i,
  MacOSMonterey: /mac os x 12[._]/i,
  MacOSVentura: /mac os x 13[._]/i,
  MacOSSonoma: /mac os x 14[._]/i,
  MacOSSequoia: /mac os x 15[._]/i,
  Mac: /os x|mac os/i,
  Linux: /linux/i,
  Linux64: /linux x86_64/i,
  ChromeOS: /cros/i,
  Wii: /wii/i,
  PS3: /playstation 3/i,
  PSP: /playstation portable/i,
  iPad: /\(iPad.*os (\d+)[._](\d+)/i,
  iPhone: /\(iPhone.*os (\d+)[._](\d+)/i,
  Bada: /Bada\/(\d+)\.(\d+)/i,
  Curl: /curl\/(\d+)\.(\d+)\.(\d+)/i
};
var PLATFORM_PATTERNS = {
  Windows: /windows nt/i,
  WindowsPhone: /windows phone/i,
  Mac: /macintosh/i,
  Linux: /linux/i,
  Wii: /wii/i,
  Playstation: /playstation/i,
  iPad: /ipad/i,
  iPod: /ipod/i,
  iPhone: /iphone/i,
  Android: /android/i,
  Blackberry: /blackberry/i,
  Samsung: /samsung/i,
  Curl: /curl/i
};
var DEFAULT_OPTIONS = {
  version: false,
  language: false,
  platform: true,
  os: true,
  pixelDepth: true,
  colorDepth: true,
  resolution: false,
  isAuthoritative: true,
  silkAccelerated: true,
  isKindleFire: true,
  isDesktop: true,
  isMobile: true,
  isTablet: true,
  isWindows: true,
  isLinux: true,
  isLinux64: true,
  isChromeOS: true,
  isMac: true,
  isiPad: true,
  isiPhone: true,
  isiPod: true,
  isAndroid: true,
  isSamsung: true,
  isSmartTV: true,
  isRaspberry: true,
  isBlackberry: true,
  isTouchScreen: true,
  isOpera: false,
  isIE: false,
  isEdge: false,
  isIECompatibilityMode: false,
  isSafari: false,
  isFirefox: false,
  isWebkit: false,
  isChrome: false,
  isKonqueror: false,
  isOmniWeb: false,
  isSeaMonkey: false,
  isFlock: false,
  isAmaya: false,
  isPhantomJS: false,
  isEpiphany: false,
  source: false,
  cpuCores: false
};
var DEFAULT_AGENT = {
  isAuthoritative: true,
  isMobile: false,
  isTablet: false,
  isiPad: false,
  isiPod: false,
  isiPhone: false,
  isAndroid: false,
  isBlackberry: false,
  isOpera: false,
  isIE: false,
  isEdge: false,
  isIECompatibilityMode: false,
  isSafari: false,
  isFirefox: false,
  isWebkit: false,
  isChrome: false,
  isKonqueror: false,
  isOmniWeb: false,
  isSeaMonkey: false,
  isFlock: false,
  isAmaya: false,
  isPhantomJS: false,
  isEpiphany: false,
  isDesktop: false,
  isWindows: false,
  isLinux: false,
  isLinux64: false,
  isMac: false,
  isChromeOS: false,
  isBada: false,
  isSamsung: false,
  isRaspberry: false,
  isBot: false,
  isCurl: false,
  isAndroidTablet: false,
  isWinJs: false,
  isKindleFire: false,
  isSilk: false,
  isCaptive: false,
  isSmartTV: false,
  isUC: false,
  isTouchScreen: false,
  silkAccelerated: false,
  colorDepth: -1,
  pixelDepth: -1,
  resolution: [0, 0],
  cpuCores: -1,
  language: "unknown",
  browser: "unknown",
  version: "unknown",
  os: "unknown",
  platform: "unknown",
  geoIp: {},
  source: ""
};

// src/utils/md5.ts
var rotateLeft = (value, shiftBits) => {
  return value << shiftBits | value >>> 32 - shiftBits;
};
var addUnsigned = (x, y) => {
  const x8 = x & 2147483648;
  const y8 = y & 2147483648;
  const x4 = x & 1073741824;
  const y4 = y & 1073741824;
  const result = (x & 1073741823) + (y & 1073741823);
  if (x4 & y4) {
    return result ^ 2147483648 ^ x8 ^ y8;
  }
  if (x4 | y4) {
    if (result & 1073741824) {
      return result ^ 3221225472 ^ x8 ^ y8;
    } else {
      return result ^ 1073741824 ^ x8 ^ y8;
    }
  } else {
    return result ^ x8 ^ y8;
  }
};
var f = (x, y, z) => {
  return x & y | ~x & z;
};
var g = (x, y, z) => {
  return x & z | y & ~z;
};
var h = (x, y, z) => {
  return x ^ y ^ z;
};
var i = (x, y, z) => {
  return y ^ (x | ~z);
};
var ff = (a, b, c, d, x, s, ac) => {
  a = addUnsigned(a, addUnsigned(addUnsigned(f(b, c, d), x), ac));
  return addUnsigned(rotateLeft(a, s), b);
};
var gg = (a, b, c, d, x, s, ac) => {
  a = addUnsigned(a, addUnsigned(addUnsigned(g(b, c, d), x), ac));
  return addUnsigned(rotateLeft(a, s), b);
};
var hh = (a, b, c, d, x, s, ac) => {
  a = addUnsigned(a, addUnsigned(addUnsigned(h(b, c, d), x), ac));
  return addUnsigned(rotateLeft(a, s), b);
};
var ii = (a, b, c, d, x, s, ac) => {
  a = addUnsigned(a, addUnsigned(addUnsigned(i(b, c, d), x), ac));
  return addUnsigned(rotateLeft(a, s), b);
};
var convertToWordArray = (str) => {
  const messageLength = str.length;
  const numberOfWordsTemp1 = messageLength + 8;
  const numberOfWordsTemp2 = (numberOfWordsTemp1 - numberOfWordsTemp1 % 64) / 64;
  const numberOfWords = (numberOfWordsTemp2 + 1) * 16;
  const wordArray = new Array(numberOfWords - 1);
  let bytePosition = 0;
  let byteCount = 0;
  while (byteCount < messageLength) {
    const wordCount2 = (byteCount - byteCount % 4) / 4;
    bytePosition = byteCount % 4 * 8;
    wordArray[wordCount2] = wordArray[wordCount2] | str.charCodeAt(byteCount) << bytePosition;
    byteCount++;
  }
  const wordCount = (byteCount - byteCount % 4) / 4;
  bytePosition = byteCount % 4 * 8;
  wordArray[wordCount] = wordArray[wordCount] | 128 << bytePosition;
  wordArray[numberOfWords - 2] = messageLength << 3;
  wordArray[numberOfWords - 1] = messageLength >>> 29;
  return wordArray;
};
var wordToHex = (value) => {
  let wordToHexValue = "";
  let wordToHexValueTemp = "";
  for (let count = 0; count <= 3; count++) {
    const byte = value >>> count * 8 & 255;
    wordToHexValueTemp = "0" + byte.toString(16);
    wordToHexValue = wordToHexValue + wordToHexValueTemp.substr(wordToHexValueTemp.length - 2, 2);
  }
  return wordToHexValue;
};
var utf8Encode = (str) => {
  str = str.replace(/\r\n/g, "\n");
  let utftext = "";
  for (let n = 0; n < str.length; n++) {
    const c = str.charCodeAt(n);
    if (c < 128) {
      utftext += String.fromCharCode(c);
    } else if (c > 127 && c < 2048) {
      utftext += String.fromCharCode(c >> 6 | 192);
      utftext += String.fromCharCode(c & 63 | 128);
    } else {
      utftext += String.fromCharCode(c >> 12 | 224);
      utftext += String.fromCharCode(c >> 6 & 63 | 128);
      utftext += String.fromCharCode(c & 63 | 128);
    }
  }
  return utftext;
};
var hashMD5 = (str) => {
  const S11 = 7, S12 = 12, S13 = 17, S14 = 22;
  const S21 = 5, S22 = 9, S23 = 14, S24 = 20;
  const S31 = 4, S32 = 11, S33 = 16, S34 = 23;
  const S41 = 6, S42 = 10, S43 = 15, S44 = 21;
  str = utf8Encode(str);
  const x = convertToWordArray(str);
  let a = 1732584193;
  let b = 4023233417;
  let c = 2562383102;
  let d = 271733878;
  for (let k = 0; k < x.length; k += 16) {
    const AA = a;
    const BB = b;
    const CC = c;
    const DD = d;
    a = ff(a, b, c, d, x[k + 0], S11, 3614090360);
    d = ff(d, a, b, c, x[k + 1], S12, 3905402710);
    c = ff(c, d, a, b, x[k + 2], S13, 606105819);
    b = ff(b, c, d, a, x[k + 3], S14, 3250441966);
    a = ff(a, b, c, d, x[k + 4], S11, 4118548399);
    d = ff(d, a, b, c, x[k + 5], S12, 1200080426);
    c = ff(c, d, a, b, x[k + 6], S13, 2821735955);
    b = ff(b, c, d, a, x[k + 7], S14, 4249261313);
    a = ff(a, b, c, d, x[k + 8], S11, 1770035416);
    d = ff(d, a, b, c, x[k + 9], S12, 2336552879);
    c = ff(c, d, a, b, x[k + 10], S13, 4294925233);
    b = ff(b, c, d, a, x[k + 11], S14, 2304563134);
    a = ff(a, b, c, d, x[k + 12], S11, 1804603682);
    d = ff(d, a, b, c, x[k + 13], S12, 4254626195);
    c = ff(c, d, a, b, x[k + 14], S13, 2792965006);
    b = ff(b, c, d, a, x[k + 15], S14, 1236535329);
    a = gg(a, b, c, d, x[k + 1], S21, 4129170786);
    d = gg(d, a, b, c, x[k + 6], S22, 3225465664);
    c = gg(c, d, a, b, x[k + 11], S23, 643717713);
    b = gg(b, c, d, a, x[k + 0], S24, 3921069994);
    a = gg(a, b, c, d, x[k + 5], S21, 3593408605);
    d = gg(d, a, b, c, x[k + 10], S22, 38016083);
    c = gg(c, d, a, b, x[k + 15], S23, 3634488961);
    b = gg(b, c, d, a, x[k + 4], S24, 3889429448);
    a = gg(a, b, c, d, x[k + 9], S21, 568446438);
    d = gg(d, a, b, c, x[k + 14], S22, 3275163606);
    c = gg(c, d, a, b, x[k + 3], S23, 4107603335);
    b = gg(b, c, d, a, x[k + 8], S24, 1163531501);
    a = gg(a, b, c, d, x[k + 13], S21, 2850285829);
    d = gg(d, a, b, c, x[k + 2], S22, 4243563512);
    c = gg(c, d, a, b, x[k + 7], S23, 1735328473);
    b = gg(b, c, d, a, x[k + 12], S24, 2368359562);
    a = hh(a, b, c, d, x[k + 5], S31, 4294588738);
    d = hh(d, a, b, c, x[k + 8], S32, 2272392833);
    c = hh(c, d, a, b, x[k + 11], S33, 1839030562);
    b = hh(b, c, d, a, x[k + 14], S34, 4259657740);
    a = hh(a, b, c, d, x[k + 1], S31, 2763975236);
    d = hh(d, a, b, c, x[k + 4], S32, 1272893353);
    c = hh(c, d, a, b, x[k + 7], S33, 4139469664);
    b = hh(b, c, d, a, x[k + 10], S34, 3200236656);
    a = hh(a, b, c, d, x[k + 13], S31, 681279174);
    d = hh(d, a, b, c, x[k + 0], S32, 3936430074);
    c = hh(c, d, a, b, x[k + 3], S33, 3572445317);
    b = hh(b, c, d, a, x[k + 6], S34, 76029189);
    a = hh(a, b, c, d, x[k + 9], S31, 3654602809);
    d = hh(d, a, b, c, x[k + 12], S32, 3873151461);
    c = hh(c, d, a, b, x[k + 15], S33, 530742520);
    b = hh(b, c, d, a, x[k + 2], S34, 3299628645);
    a = ii(a, b, c, d, x[k + 0], S41, 4096336452);
    d = ii(d, a, b, c, x[k + 7], S42, 1126891415);
    c = ii(c, d, a, b, x[k + 14], S43, 2878612391);
    b = ii(b, c, d, a, x[k + 5], S44, 4237533241);
    a = ii(a, b, c, d, x[k + 12], S41, 1700485571);
    d = ii(d, a, b, c, x[k + 3], S42, 2399980690);
    c = ii(c, d, a, b, x[k + 10], S43, 4293915773);
    b = ii(b, c, d, a, x[k + 1], S44, 2240044497);
    a = ii(a, b, c, d, x[k + 8], S41, 1873313359);
    d = ii(d, a, b, c, x[k + 15], S42, 4264355552);
    c = ii(c, d, a, b, x[k + 6], S43, 2734768916);
    b = ii(b, c, d, a, x[k + 13], S44, 1309151649);
    a = ii(a, b, c, d, x[k + 4], S41, 4149444226);
    d = ii(d, a, b, c, x[k + 11], S42, 3174756917);
    c = ii(c, d, a, b, x[k + 2], S43, 718787259);
    b = ii(b, c, d, a, x[k + 9], S44, 3951481745);
    a = addUnsigned(a, AA);
    b = addUnsigned(b, BB);
    c = addUnsigned(c, CC);
    d = addUnsigned(d, DD);
  }
  const temp = wordToHex(a) + wordToHex(b) + wordToHex(c) + wordToHex(d);
  return temp.toLowerCase();
};
var hashInt = (str) => {
  let hash = 0;
  if (str.length === 0) return hash;
  for (let i2 = 0; i2 < str.length; i2++) {
    const chr = str.charCodeAt(i2);
    hash = (hash << 5) - hash + chr;
    hash |= 0;
  }
  return hash;
};

// src/utils/environment.ts
var isBrowser = () => {
  return typeof window !== "undefined" && typeof window.document !== "undefined";
};
var getNavigator = () => {
  if (isBrowser() && typeof navigator !== "undefined") {
    return navigator;
  }
  return void 0;
};
var getScreen = () => {
  if (isBrowser() && typeof screen !== "undefined") {
    return screen;
  }
  return void 0;
};
var getWindow = () => {
  if (isBrowser() && typeof window !== "undefined") {
    return window;
  }
  return void 0;
};
var getUserAgent = () => {
  const nav = getNavigator();
  return nav?.userAgent || "";
};
var getLanguage = () => {
  const nav = getNavigator();
  if (!nav) return "unknown";
  return (nav.language || nav.userLanguage || nav.browserLanguage || nav.systemLanguage || "").toLowerCase() || "unknown";
};
var getColorDepth = () => {
  const scr = getScreen();
  return scr?.colorDepth ?? -1;
};
var getPixelDepth = () => {
  const scr = getScreen();
  return scr?.pixelDepth ?? -1;
};
var getScreenResolution = () => {
  const scr = getScreen();
  if (scr) {
    return [scr.availWidth || 0, scr.availHeight || 0];
  }
  return [0, 0];
};
var getCPUCores = () => {
  const nav = getNavigator();
  return nav?.hardwareConcurrency ?? -1;
};
var isTouchScreen = () => {
  const win = getWindow();
  const nav = getNavigator();
  if (!win || !nav) return false;
  return "ontouchstart" in win || nav.maxTouchPoints !== void 0 && nav.maxTouchPoints > 0 || nav.msMaxTouchPoints !== void 0 && nav.msMaxTouchPoints > 0;
};

// src/core/DeviceUUID.ts
var DeviceUUID = class _DeviceUUID {
  /**
   * Create a new DeviceUUID instance
   * @param options - Configuration options
   */
  constructor(options = {}) {
    this.versionPatterns = VERSION_PATTERNS;
    this.browserPatterns = BROWSER_PATTERNS;
    this.osPatterns = OS_PATTERNS;
    this.platformPatterns = PLATFORM_PATTERNS;
    this.options = { ...DEFAULT_OPTIONS, ...options };
    this.agent = {
      ...DEFAULT_AGENT,
      hashInt,
      hashMD5
    };
  }
  /**
   * Get browser name from user agent string
   */
  getBrowser(source) {
    if (this.browserPatterns.Edge.test(source)) {
      this.agent.isEdge = true;
      return "Edge";
    }
    if (this.browserPatterns.PhantomJS.test(source)) {
      this.agent.isPhantomJS = true;
      return "PhantomJS";
    }
    if (this.browserPatterns.Konqueror.test(source)) {
      this.agent.isKonqueror = true;
      return "Konqueror";
    }
    if (this.browserPatterns.Amaya.test(source)) {
      this.agent.isAmaya = true;
      return "Amaya";
    }
    if (this.browserPatterns.Epiphany.test(source)) {
      this.agent.isEpiphany = true;
      return "Epiphany";
    }
    if (this.browserPatterns.SeaMonkey.test(source)) {
      this.agent.isSeaMonkey = true;
      return "SeaMonkey";
    }
    if (this.browserPatterns.Flock.test(source)) {
      this.agent.isFlock = true;
      return "Flock";
    }
    if (this.browserPatterns.OmniWeb.test(source)) {
      this.agent.isOmniWeb = true;
      return "OmniWeb";
    }
    if (this.browserPatterns.Opera.test(source)) {
      this.agent.isOpera = true;
      return "Opera";
    }
    if (this.browserPatterns.Chromium.test(source)) {
      this.agent.isChrome = true;
      return "Chromium";
    }
    if (this.browserPatterns.Chrome.test(source)) {
      this.agent.isChrome = true;
      return "Chrome";
    }
    if (this.browserPatterns.Safari.test(source)) {
      this.agent.isSafari = true;
      return "Safari";
    }
    if (this.browserPatterns.WinJs.test(source)) {
      this.agent.isWinJs = true;
      return "WinJs";
    }
    if (this.browserPatterns.IE.test(source)) {
      this.agent.isIE = true;
      return "IE";
    }
    if (this.browserPatterns.PS3.test(source)) {
      return "ps3";
    }
    if (this.browserPatterns.PSP.test(source)) {
      return "psp";
    }
    if (this.browserPatterns.Firefox.test(source)) {
      this.agent.isFirefox = true;
      return "Firefox";
    }
    if (this.browserPatterns.UC.test(source)) {
      this.agent.isUC = true;
      return "UCBrowser";
    }
    if (source.indexOf("Mozilla") !== 0 && /^([\d\w-.]+)\/[\d\w.-]+/i.test(source)) {
      this.agent.isAuthoritative = false;
      return RegExp.$1;
    }
    return "unknown";
  }
  /**
   * Get browser version from user agent string
   */
  getBrowserVersion(source) {
    const browser = this.agent.browser;
    const versionMap = {
      Edge: this.versionPatterns.Edge,
      PhantomJS: this.versionPatterns.PhantomJS,
      Chrome: this.versionPatterns.Chrome,
      Chromium: this.versionPatterns.Chromium,
      Safari: this.versionPatterns.Safari,
      Opera: this.versionPatterns.Opera,
      Firefox: this.versionPatterns.Firefox,
      WinJs: this.versionPatterns.WinJs,
      IE: this.versionPatterns.IE,
      ps3: this.versionPatterns.Ps3,
      psp: this.versionPatterns.Psp,
      Amaya: this.versionPatterns.Amaya,
      Epiphany: this.versionPatterns.Epiphany,
      SeaMonkey: this.versionPatterns.SeaMonkey,
      Flock: this.versionPatterns.Flock,
      OmniWeb: this.versionPatterns.OmniWeb,
      UCBrowser: this.versionPatterns.UC
    };
    const pattern = versionMap[browser];
    if (pattern && pattern.test(source)) {
      if (browser === "IE" || browser === "Opera") {
        return RegExp.$2 || RegExp.$1 || "unknown";
      }
      return RegExp.$1 || "unknown";
    }
    if (browser !== "unknown") {
      const regex = new RegExp(`${browser}[\\/ ]([\\d\\w.\\-]+)`, "i");
      if (regex.test(source)) {
        return RegExp.$1 || "unknown";
      }
    }
    return "unknown";
  }
  /**
   * Get operating system from user agent string
   */
  getOS(source) {
    if (this.osPatterns.Windows11.test(source)) {
      const match = source.match(this.osPatterns.Windows11);
      if (match && match[1]) {
        const version = parseInt(match[1]);
        if (version >= 96) {
          this.agent.isWindows = true;
          return "Windows 11";
        }
      }
    }
    if (this.osPatterns.Windows10.test(source)) {
      this.agent.isWindows = true;
      return "Windows 10.0";
    }
    if (this.osPatterns.WindowsVista.test(source)) {
      this.agent.isWindows = true;
      return "Windows Vista";
    }
    if (this.osPatterns.Windows7.test(source)) {
      this.agent.isWindows = true;
      return "Windows 7";
    }
    if (this.osPatterns.Windows8.test(source)) {
      this.agent.isWindows = true;
      return "Windows 8";
    }
    if (this.osPatterns.Windows81.test(source)) {
      this.agent.isWindows = true;
      return "Windows 8.1";
    }
    if (this.osPatterns.Windows2003.test(source)) {
      this.agent.isWindows = true;
      return "Windows 2003";
    }
    if (this.osPatterns.WindowsXP.test(source)) {
      this.agent.isWindows = true;
      return "Windows XP";
    }
    if (this.osPatterns.Windows2000.test(source)) {
      this.agent.isWindows = true;
      return "Windows 2000";
    }
    if (this.osPatterns.WindowsPhone8.test(source)) {
      return "Windows Phone 8";
    }
    if (this.osPatterns.Linux64.test(source)) {
      this.agent.isLinux = true;
      this.agent.isLinux64 = true;
      return "Linux 64";
    }
    if (this.osPatterns.Linux.test(source)) {
      this.agent.isLinux = true;
      return "Linux";
    }
    if (this.osPatterns.ChromeOS.test(source)) {
      this.agent.isChromeOS = true;
      return "Chrome OS";
    }
    if (this.osPatterns.Wii.test(source)) {
      return "Wii";
    }
    if (this.osPatterns.PS3.test(source)) {
      return "Playstation";
    }
    if (this.osPatterns.PSP.test(source)) {
      return "Playstation";
    }
    if (this.osPatterns.OSXCheetah.test(source)) {
      this.agent.isMac = true;
      return "OS X Cheetah";
    }
    if (this.osPatterns.OSXPuma.test(source)) {
      this.agent.isMac = true;
      return "OS X Puma";
    }
    if (this.osPatterns.OSXJaguar.test(source)) {
      this.agent.isMac = true;
      return "OS X Jaguar";
    }
    if (this.osPatterns.OSXPanther.test(source)) {
      this.agent.isMac = true;
      return "OS X Panther";
    }
    if (this.osPatterns.OSXTiger.test(source)) {
      this.agent.isMac = true;
      return "OS X Tiger";
    }
    if (this.osPatterns.OSXLeopard.test(source)) {
      this.agent.isMac = true;
      return "OS X Leopard";
    }
    if (this.osPatterns.OSXSnowLeopard.test(source)) {
      this.agent.isMac = true;
      return "OS X Snow Leopard";
    }
    if (this.osPatterns.OSXLion.test(source)) {
      this.agent.isMac = true;
      return "OS X Lion";
    }
    if (this.osPatterns.OSXMountainLion.test(source)) {
      this.agent.isMac = true;
      return "OS X Mountain Lion";
    }
    if (this.osPatterns.OSXMavericks.test(source)) {
      this.agent.isMac = true;
      return "OS X Mavericks";
    }
    if (this.osPatterns.OSXYosemite.test(source)) {
      this.agent.isMac = true;
      return "OS X Yosemite";
    }
    if (this.osPatterns.OSXElCapitan.test(source)) {
      this.agent.isMac = true;
      return "OS X El Capitan";
    }
    if (this.osPatterns.OSXSierra.test(source)) {
      this.agent.isMac = true;
      return "macOS Sierra";
    }
    if (this.osPatterns.OSXHighSierra.test(source)) {
      this.agent.isMac = true;
      return "macOS High Sierra";
    }
    if (this.osPatterns.OSXMojave.test(source)) {
      this.agent.isMac = true;
      return "macOS Mojave";
    }
    if (this.osPatterns.OSXCatalina.test(source)) {
      this.agent.isMac = true;
      return "macOS Catalina";
    }
    if (this.osPatterns.MacOSBigSur.test(source)) {
      this.agent.isMac = true;
      return "macOS Big Sur";
    }
    if (this.osPatterns.MacOSMonterey.test(source)) {
      this.agent.isMac = true;
      return "macOS Monterey";
    }
    if (this.osPatterns.MacOSVentura.test(source)) {
      this.agent.isMac = true;
      return "macOS Ventura";
    }
    if (this.osPatterns.MacOSSonoma.test(source)) {
      this.agent.isMac = true;
      return "macOS Sonoma";
    }
    if (this.osPatterns.MacOSSequoia.test(source)) {
      this.agent.isMac = true;
      return "macOS Sequoia";
    }
    if (this.osPatterns.iPad.test(source)) {
      this.agent.isiPad = true;
      return "iOS";
    }
    if (this.osPatterns.iPhone.test(source)) {
      this.agent.isiPhone = true;
      return "iOS";
    }
    if (this.osPatterns.Mac.test(source)) {
      this.agent.isMac = true;
      return "Mac OS";
    }
    if (this.osPatterns.Bada.test(source)) {
      this.agent.isBada = true;
      return "Bada";
    }
    if (this.osPatterns.Curl.test(source)) {
      this.agent.isCurl = true;
      return "Curl";
    }
    return "unknown";
  }
  /**
   * Get platform from user agent string
   */
  getPlatform(source) {
    if (this.platformPatterns.Windows.test(source)) {
      return "Microsoft Windows";
    }
    if (this.platformPatterns.WindowsPhone.test(source)) {
      this.agent.isWindowsPhone = true;
      return "Microsoft Windows Phone";
    }
    if (this.platformPatterns.Mac.test(source)) {
      return "Apple Mac";
    }
    if (this.platformPatterns.Curl.test(source)) {
      return "Curl";
    }
    if (this.platformPatterns.Android.test(source)) {
      this.agent.isAndroid = true;
      return "Android";
    }
    if (this.platformPatterns.Blackberry.test(source)) {
      this.agent.isBlackberry = true;
      return "Blackberry";
    }
    if (this.platformPatterns.Linux.test(source)) {
      return "Linux";
    }
    if (this.platformPatterns.Wii.test(source)) {
      return "Wii";
    }
    if (this.platformPatterns.Playstation.test(source)) {
      return "Playstation";
    }
    if (this.platformPatterns.iPad.test(source)) {
      this.agent.isiPad = true;
      return "iPad";
    }
    if (this.platformPatterns.iPod.test(source)) {
      this.agent.isiPod = true;
      return "iPod";
    }
    if (this.platformPatterns.iPhone.test(source)) {
      this.agent.isiPhone = true;
      return "iPhone";
    }
    if (this.platformPatterns.Samsung.test(source)) {
      this.agent.isSamsung = true;
      return "Samsung";
    }
    return "unknown";
  }
  /**
   * Test for bot/crawler
   */
  testBot() {
    const isBot = IS_BOT_REGEXP.exec(this.agent.source.toLowerCase());
    if (isBot) {
      this.agent.isBot = isBot[1];
    } else if (!this.agent.isAuthoritative) {
      this.agent.isBot = /bot/i.test(this.agent.source);
    }
  }
  /**
   * Test for Smart TV
   */
  testSmartTV() {
    const isSmartTV = /smart-tv|smarttv|googletv|appletv|hbbtv|pov_tv|netcast.tv/gi.exec(
      this.agent.source.toLowerCase()
    );
    if (isSmartTV) {
      this.agent.isSmartTV = isSmartTV[1] || true;
    }
  }
  /**
   * Test for mobile device
   */
  testMobile() {
    if (this.agent.isSmartTV) {
      this.agent.isMobile = false;
      this.agent.isDesktop = false;
      return;
    }
    if (this.agent.isAndroid || this.agent.isSamsung || this.agent.isiPhone || this.agent.isiPod || this.agent.isBada || this.agent.isBlackberry || this.agent.isWindowsPhone) {
      this.agent.isMobile = true;
      this.agent.isDesktop = false;
      return;
    }
    if (this.agent.isiPad) {
      this.agent.isMobile = false;
      this.agent.isDesktop = false;
      return;
    }
    if (this.agent.isWindows || this.agent.isLinux || this.agent.isMac || this.agent.isChromeOS) {
      this.agent.isDesktop = true;
      return;
    }
    if (/mobile/i.test(this.agent.source)) {
      this.agent.isMobile = true;
      this.agent.isDesktop = false;
    }
  }
  /**
   * Test for Android tablet
   */
  testAndroidTablet() {
    if (this.agent.isAndroid && !/mobile/i.test(this.agent.source)) {
      this.agent.isAndroidTablet = true;
      this.agent.isMobile = false;
    }
  }
  /**
   * Test for tablet device
   */
  testTablet() {
    if (this.agent.isiPad || this.agent.isAndroidTablet || this.agent.isKindleFire) {
      this.agent.isTablet = true;
    }
    if (/tablet/i.test(this.agent.source)) {
      this.agent.isTablet = true;
    }
  }
  /**
   * Test for IE compatibility mode
   */
  testCompatibilityMode() {
    if (this.agent.isIE) {
      const tridentMatch = /Trident\/(\d)\.0/i.exec(this.agent.source);
      if (tridentMatch) {
        const tridentVersion = parseInt(tridentMatch[1], 10);
        const version = parseFloat(this.agent.version);
        if (version === 7 && tridentVersion === 7) {
          this.agent.isIECompatibilityMode = true;
          this.agent.version = "11.0";
        } else if (version === 7 && tridentVersion === 6) {
          this.agent.isIECompatibilityMode = true;
          this.agent.version = "10.0";
        } else if (version === 7 && tridentVersion === 5) {
          this.agent.isIECompatibilityMode = true;
          this.agent.version = "9.0";
        } else if (version === 7 && tridentVersion === 4) {
          this.agent.isIECompatibilityMode = true;
          this.agent.version = "8.0";
        }
      }
    }
  }
  /**
   * Test for Amazon Silk browser
   */
  testSilk() {
    if (/silk/gi.test(this.agent.source)) {
      this.agent.isSilk = true;
    }
    if (/Silk-Accelerated=true/gi.test(this.agent.source)) {
      this.agent.silkAccelerated = true;
    }
  }
  /**
   * Test for Kindle Fire device
   */
  testKindleFire() {
    const kindleTests = [
      [/KFOT/gi, "Kindle Fire"],
      [/KFTT/gi, "Kindle Fire HD"],
      [/KFJWI/gi, "Kindle Fire HD 8.9"],
      [/KFJWA/gi, "Kindle Fire HD 8.9 4G"],
      [/KFSOWI/gi, "Kindle Fire HD 7"],
      [/KFTHWI/gi, "Kindle Fire HDX 7"],
      [/KFTHWA/gi, "Kindle Fire HDX 7 4G"],
      [/KFAPWI/gi, "Kindle Fire HDX 8.9"],
      [/KFAPWA/gi, "Kindle Fire HDX 8.9 4G"],
      [/KFMAWI/gi, "Kindle Fire HD 10"]
    ];
    for (const [pattern] of kindleTests) {
      if (pattern.test(this.agent.source)) {
        this.agent.isKindleFire = true;
        return;
      }
    }
  }
  /**
   * Test for Captive Network Assistant
   */
  testCaptiveNetwork() {
    if (/CaptiveNetwork/gi.test(this.agent.source)) {
      this.agent.isCaptive = true;
      this.agent.isMac = true;
      this.agent.platform = "Apple Mac";
    }
  }
  /**
   * Test for touch screen support
   */
  testTouchSupport() {
    this.agent.isTouchScreen = isTouchScreen();
  }
  /**
   * Get language from browser
   */
  getLanguageInfo() {
    this.agent.language = getLanguage();
  }
  /**
   * Get color depth
   */
  getColorDepthInfo() {
    this.agent.colorDepth = getColorDepth();
  }
  /**
   * Get pixel depth
   */
  getPixelDepthInfo() {
    this.agent.pixelDepth = getPixelDepth();
  }
  /**
   * Get screen resolution
   */
  getScreenResolutionInfo() {
    this.agent.resolution = getScreenResolution();
  }
  /**
   * Get CPU core count
   */
  getCPUInfo() {
    this.agent.cpuCores = getCPUCores();
  }
  /**
   * Reset agent to default state
   */
  reset() {
    this.agent = {
      ...DEFAULT_AGENT,
      hashInt,
      hashMD5
    };
    return this;
  }
  /**
   * Parse user agent and collect device information
   * @param source - User agent string (defaults to navigator.userAgent)
   * @returns AgentInfo object with device details
   */
  parse(source) {
    const ua = new _DeviceUUID();
    const userAgent = source || getUserAgent();
    ua.agent.source = userAgent.replace(/^\s*/, "").replace(/\s*$/, "");
    ua.agent.os = ua.getOS(ua.agent.source);
    ua.agent.platform = ua.getPlatform(ua.agent.source);
    ua.agent.browser = ua.getBrowser(ua.agent.source);
    ua.agent.version = ua.getBrowserVersion(ua.agent.source);
    ua.testBot();
    ua.testSmartTV();
    ua.testMobile();
    ua.testAndroidTablet();
    ua.testTablet();
    ua.testCompatibilityMode();
    ua.testSilk();
    ua.testKindleFire();
    ua.testCaptiveNetwork();
    ua.testTouchSupport();
    ua.getLanguageInfo();
    ua.getColorDepthInfo();
    ua.getPixelDepthInfo();
    ua.getScreenResolutionInfo();
    ua.getCPUInfo();
    return ua.agent;
  }
  /**
   * Generate a UUID based on device characteristics
   * @param customData - Optional custom data to include in UUID generation
   * @returns UUID string in v4 format
   */
  get(customData) {
    const du = this.parse();
    const dataArray = [];
    for (const key in this.options) {
      if (Object.prototype.hasOwnProperty.call(this.options, key)) {
        const value = du[key];
        dataArray.push(value);
      }
    }
    if (customData) {
      dataArray.push(customData);
    }
    if (!this.options.resolution && du.isMobile) {
      dataArray.push(du.resolution);
    }
    const pref = "b";
    const tmpUuid = hashMD5(dataArray.join(":"));
    const uuid = [
      tmpUuid.slice(0, 8),
      tmpUuid.slice(8, 12),
      "4" + tmpUuid.slice(12, 15),
      // Version 4
      pref + tmpUuid.slice(15, 18),
      // Variant bits
      tmpUuid.slice(20)
    ];
    return uuid.join("-");
  }
};
/**
 * device-uuid v2.0.0
 * Fast browser device UUID generation library
 *
 * @author Alexey Gordeyev
 * @license MIT
 */

exports.BOTS = BOTS;
exports.BROWSER_PATTERNS = BROWSER_PATTERNS;
exports.DEFAULT_OPTIONS = DEFAULT_OPTIONS;
exports.DeviceUUID = DeviceUUID;
exports.OS_PATTERNS = OS_PATTERNS;
exports.PLATFORM_PATTERNS = PLATFORM_PATTERNS;
exports.VERSION_PATTERNS = VERSION_PATTERNS;
exports.default = DeviceUUID;
exports.hashInt = hashInt;
exports.hashMD5 = hashMD5;
